### PR TITLE
feat(errors): consolidate multiple sarif reports into a single report

### PR DESCRIPTION
**Pull request by royalpinto007**
Consolidates multiple SARIF reports into a single report.

### CODE REVIEW

File: compiler/src/dmd/errors.d [hunk 1]
```diff
@@ -37,6 +37,23 @@ enum ErrorKind
...
+  kind = The type of diagnostic, indicating whether it is an error, warning, deprecation, etc.
+*/
+struct Diagnostic
+{
+    SourceLoc loc;
+    string message;
+    ErrorKind kind;
+}
+
+Diagnostic[] diagnostics; // Global array to store diagnostics
```
Comment (thewilsonator) [inline compiler/src/dmd/errors.d:55]: this is thread local, which it shouldn't be. make it `__gshared`.  also it should be set to empty with https://github.com/dlang/dmd/blob/master/compiler/src/dmd/frontend.d#L172

File: compiler/src/dmd/errors.d [hunk 2]
```diff
@@ -101,6 +118,153 @@ class ErrorSinkCompiler : ErrorSink
         verrorReport(loc, format, ap, ErrorKind.message);
         va_end(ap);
     }
+
+    void plugSink()
+    {
+        // Only proceed if there are collected diagnostics
+        if (diagnostics.length > 0)
```
Comment (thewilsonator) [inline compiler/src/dmd/errors.d:125]:
```
if (!diagnostics.length)
    return;
...
```

File: compiler/src/dmd/errors.d [hunk 3]
```diff
@@ -476,7 +648,7 @@ private extern(C++) void verrorReport(const SourceLoc loc, const(char)* format,
             info.headerColor = Classification.error;
             if (global.params.v.messageStyle == MessageStyle.sarif)
             {
-                generateSarifReport(loc, format, ap, info.kind, false);
+                diagnostics ~= Diagnostic(loc, formattedMessage, kind);
```
Comment (thewilsonator) [inline compiler/src/dmd/errors.d:651]: just modify `generateSarifReport` report to do this, and make the formatting changes there

**Comment by royalpinto007** [comment]
@thewilsonator, can you please help me understand why the CI is failing? 
The logs aren’t giving me much information.

**Review by thewilsonator — APPROVED** [review]

**Pull request merged**